adr/0005-code-formatting.md

```diff
@@ -0,0 +1,31 @@
+# 5. Code Formatting
+
+Date: 2019-07-07
+
+Architecture issue: [#249](https://github.com/home-assistant/architecture/issues/249)
+
+## Status
+
+Accepted
+
+## Context
+
+Proper code formatting is vital to the long-term viability and maintainability of a code
+base. In place, a codebase with such formatting is:
+
+- easier to review
+- easier to diff against previous commits
+- more likely to be free of linting errors
+
+## Proposal
+
+- We use [Black](https://github.com/python/black) as code formatter for the back-end codebase.
+- We use Black's default options.
+- We utilize a bot (such as [the black_out bot](https://github.com/Mariatta/black_out)) to ensure PRs are appropriately formatted before being accepted.
+- We recommend and document the set-up of Black, `pylint`, and `flake8` pre-commit hooks in the development environment.
```

Comment:
Are not `pylint` and `flake8` precommit hooks offtopic for the purpose of this ADR? Not that I think that they shouldn't be used, but in my view they are not relevant to the topic at hand (outside the configuration changes described below).

Reply:
Kind of.

You want code to be consistent and valid at every commit. Since this ADR has become broader than adding `black` itself (since it changes `flake8` and `pylint` settings), I thought these were relevant.

In my opinion the minimal recommendation, while being on topic, of only adding black commit hooks definitely is wrong.

I would be in favor of opening up the scope of the ADR to concern all code formatting and linting. I think that the concerns about `flake8` and `pylint` show that a slightly more holistic approach may be needed here.

Reply:
I'm in favor of adding `pre-commit` to our project, nevertheless, this should really be a separate ADR.

Reply:
Enforced coding standards _are_ already a part of the project. Both through CircleCI and the recommendation to set-up flake8 pre-commit hooks in the [development and testing](https://developers.home-assistant.io/docs/en/development_testing.html#preventing-linter-errors) instructions. I see adding removing flake8 rules and adding black without pre-commit as a regression compared to the status quo.

Potentially there is a trivial ADR on coding standards here.

Specific to this ADR, I do think you want black to be ran locally? Iteratively fixing issues after CircleCI runs fails or issue comments is noisy and imo is an anti-pattern.

Reply:
Let discuss this in the ADR issue instead of the PR.